parsec/core/gui/app.py

```diff
@@ -49,41 +49,48 @@ def _before_quit():
 
 
 async def _start_ipc_server(config, main_window, start_arg, result_queue):
-    new_instance_needed_qt = ThreadSafeQtSignal(main_window, "new_instance_needed", object)
-    foreground_needed_qt = ThreadSafeQtSignal(main_window, "foreground_needed")
-
-    async def cmd_handler(cmd):
-        if cmd["cmd"] == IPCCommand.FOREGROUND:
-            foreground_needed_qt.emit()
-        elif cmd["cmd"] == IPCCommand.NEW_INSTANCE:
-            new_instance_needed_qt.emit(cmd.get("start_arg"))
-        return {"status": "ok"}
-
-    while True:
-        try:
-            async with run_ipc_server(
-                cmd_handler, config.ipc_socket_file, win32_mutex_name=config.ipc_win32_mutex_name
-            ):
-                result_queue.put("started")
-                await trio.sleep_forever()
-
-        except IPCServerAlreadyRunning:
-            # Parsec is already started, give it our work then
+    try:
+        new_instance_needed_qt = ThreadSafeQtSignal(main_window, "new_instance_needed", object)
+        foreground_needed_qt = ThreadSafeQtSignal(main_window, "foreground_needed")
+
+        async def _cmd_handler(cmd):
+            if cmd["cmd"] == IPCCommand.FOREGROUND:
+                foreground_needed_qt.emit()
+            elif cmd["cmd"] == IPCCommand.NEW_INSTANCE:
+                new_instance_needed_qt.emit(cmd.get("start_arg"))
+            return {"status": "ok"}
+
+        while True:
             try:
+                async with run_ipc_server(
+                    _cmd_handler,
+                    config.ipc_socket_file,
+                    win32_mutex_name=config.ipc_win32_mutex_name,
+                ):
+                    result_queue.put_nowait("started")
+                    await trio.sleep_forever()
+
+            except IPCServerAlreadyRunning:
+                # Parsec is already started, give it our work then
                 try:
                     if start_arg:
                         await send_to_ipc_server(
                             config.ipc_socket_file, IPCCommand.NEW_INSTANCE, start_arg=start_arg
                         )
                     else:
                         await send_to_ipc_server(config.ipc_socket_file, IPCCommand.FOREGROUND)
-                finally:
-                    result_queue.put("already_running")
-                return
 
-            except IPCServerNotRunning:
-                # IPC server has closed, retry to create our own
-                continue
+                except IPCServerNotRunning:
+                    # IPC server has closed, retry to create our own
+                    continue
+
+                # We have successfuly noticed the other running application
+                result_queue.put_nowait("already_running")
+                break
+
+    except Exception:
+        result_queue.put_nowait("error")
+        raise
```

Comment:
What is the implication of the `raise` here? As far as I can tell the corresponding `run_ipc_server` job is not used so this exception is likely lost. The exception could also go in the queue.

More generally, the fact that we're mixing job communication and custom queue communication makes it a bit hard to read.

Reply:
>  job is not used so this exception is likely lost

trio_thread module handles uncatched exceptions of all submitted jobs:
https://github.com/Scille/parsec-cloud/blob/2d877db087a1203c3024b963af0f356a42f39ef3/parsec/core/gui/trio_thread.py#L132-L136

But I agree this is a bit hidden, I'll add a comment about this

> More generally, the fact that we're mixing job communication and custom queue communication makes it a bit hard to read.

Yes, we should rewrite this code so that `run_gui` is divided in two callbacks (one before and one after ipc server has started). However it's not trivial to do given the `app.exec_` at the very end of the function (i.e. the Qt loop is not started until then, so no signal can be handled...)
Another solution would be to start the ipc server in a thread not related to the trio_thread stuff, but it's also messy because we want to run ipc_server stuff in the trio loop (which is handled by trio_trhead)